compiler/rustc_parse/src/lexer/mod.rs

```diff
@@ -64,28 +64,34 @@ pub(crate) fn parse_token_trees<'a>(
         override_span,
         nbsp_is_whitespace: false,
     };
-    let (token_trees, unmatched_delims) =
+    let (stream, res, unmatched_delims) =
         tokentrees::TokenTreesReader::parse_all_token_trees(string_reader);
-    match token_trees {
-        Ok(stream) if unmatched_delims.is_empty() => Ok(stream),
-        _ => {
-            // Return error if there are unmatched delimiters or unclosed delimiters.
-            // We emit delimiter mismatch errors first, then emit the unclosing delimiter mismatch
-            // because the delimiter mismatch is more likely to be the root cause of error
-
-            let mut buffer = Vec::with_capacity(1);
-            for unmatched in unmatched_delims {
-                if let Some(err) = make_unclosed_delims_error(unmatched, &sess) {
-                    err.buffer(&mut buffer);
-                }
+    if !unmatched_delims.is_empty() || res.is_err() {
+        // Return error if there are unmatched delimiters or unclosed delimiters.
+        // We emit delimiter mismatch errors first, then emit the unclosing delimiter mismatch
+        // because the delimiter mismatch is more likely to be the root cause of error
+
+        let mut buffer = Vec::with_capacity(1);
+        let mut parser = crate::stream_to_parser(sess, stream.clone(), None);
+        while parser.token != token::Eof {
+            // There were unmatched delims, see if there were diff markers that caused it.
+            if let Err(err) = parser.err_diff_marker() {
+                err.buffer(&mut buffer);
```

Comment:
So the logic here is that errors are buffered for later and not emitted immediately.
But `parse_token_tree_open_delim` is eventually called from here and emits the diff marker immediately.

I think `parse_token_tree_open_delim` should return the diff marker error instead of the unclosed delimiter error, if it occurs. Then it will be buffered here normally.

Also this piece of code
```rust
let mut parser = crate::stream_to_parser(...);
while parser.token != token::Eof { ... }
```
is better factored out and reused between this place and `parse_token_tree_open_delim`.

Reply:
This part actually was unneded, because the code in parse_token_tree_open_delim was sufficient. I must have needed it during development but can't trigger it now.